## Abstract {.page_break_before}

The current reference genome is the backbone of diverse and rich annotations. 
Simple text formats, like VCF or BED, have been widely adopted and helped the critical exchange of genomic information. 
There is a dire need for tools and formats enabling pangenomic annotation to facilitate such enrichment pangenomic references.
The Graph Alignment Format (GAF) is a text format, tab-delimitted like BED/VCF files, which was proposed to represent alignments.
GAF could also be used to store paths representing annotations in a pangenome graph, but there are no tools to index and query them efficiently.

Here, we present extensions to vg and HTSlib that provide efficient sorting, indexing, and querying for GAF files.
With this approach, annotations overlapping a subgraph can be extracted quickly.
Paths are sorted based on the IDs of traversed nodes, compressed with BGZIP, and indexed with HTSlib/tabix via our extensions for the GAF format. 
Compared to the binary GAM format, text-based GAF files are easier to write, and we show that they are twice as fast to sort and half as large on disk.
In addition, we updated `vg annotate`, which takes BED or GFF3 annotation files relative to linear sequences and projects them into the pangenome. 
It can now produce GAF files representing these annotations' paths through the pangenome.

We showcase these new tools on several applications. 
We projected annotations for all Human Pangenome Reference Consortium Year 1 haplotypes, including genes, segmental duplications, tandem repeats and repeats annotations, into the Minigraph-Cactus pangenome (GRCh38-based v1.1). 
We also projected known variants from the GWAS Catalog and expression QTLs from the GTEx project into the pangenome. 
Finally, we reanalyzed ATAC-seq data from ENCODE to demonstrate what a coverage track could look like in a pangenome graph.
These rich annotations can be quickly queried with vg and visualized using existing tools like the Sequence Tube Map or Bandage.
