## Methods

### Indexing paths in GAF files

The sorting and indexing algorithm is most efficient and makes sense when node IDs are integers, sorted based on the topology of the pangenome graph.
This is the case for pangenomes constructed by minigraph-cactus[@minigraph_cactus], PGGB[@pggb], or `vg construct`[@vg].
Otherwise, the pangenome graph can be *sorted*, by changing the node IDs, using `vg` or `odgi`[@odgi].
If the node IDs are sorted integers, a short path through the graph should only traverse nodes with IDs contained in a small interval (see Figure {@fig:index}A).
The main approach of the GAF sorting and indexing approach is to work with those intervals.
Hence, to sort a GAF file, each path is first parsed to extract the minimum and maximum node IDs (see Figure {@fig:index}B).
The paths are then sorted first by their minimum ID, and then by their maximum ID.
(Nodes are assumed to be short enough that within-node indexing is not required.)
This is similar to the approach used to sort BED or VCF files on linear genomes: they are sorted by sequence name, then start position, then end position.

![
**Path sorting and indexing using vg and HTSlib/tabix.**
**A.** A region of pangenome is represented with nodes (containing sequences) in yellow and edges in black. 
The node IDs are topologically sorted integers, ranging here from 23 to 36.
Three paths are highlighted in red, blue and green. 
**B.** The three paths are written with the GAF syntax, by specifying the orientations (`<`/`>`) and IDs of the traversed nodes.
For each path, the node range *n-N*, between the minimum and maximum node IDs, is used for sorting the path.
**C.** Overview of the workflow to sort a GAF file using `vg gamsort`, compress it with `bgzip` and index using `tabix`. 
**D.** The small *.tbi* index file helps query slices of the GAF file quickly, and is usable with `tabix`, or `vg` subcommands like `find` or `chunk`.
](figures/gafindexing.png "GAF indexing"){#fig:index}

We added a GAF sorting feature to the `vg` toolkit's `gamsort` subcommand, available in `vg` version 1.56 and later (see Figure {@fig:index}C).
It first sorts chunks of GAF records (e.g. reads), to avoid having to hold the entire set in memory.
The sorted chunks are then merged into the final sorted GAF file.
A sorted GAF file can be compressed in the BGZF format (or *bgzipped*) using `bgzip`.
The BGZF format is a backwards-compatible extension of the standard gzip format that compresses a file by block to facilitate random access.
It is used to compress a number of file types popular in bioinformatics, including VCF, BED, and BAM files, and is implemented in HTSlib[@htslib].

We extended HTSlib[@htslib] to allow indexing bgzipped GAF files.
Similar to other tab-separated file like VCF or BED, a `gaf` preset was added to `tabix`.
For BED or VCF, `tabix` extract the interval information from the columns with the sequence name and genomic position[@tabix]. 
In the new `gaf` preset, it instead parses the path information in the GAF file to extract the minimum and maximum node IDs.
The indexing is then based on this interval, the same as used for the sorting described above.

We tested the GAF sorting and indexing performance on 30X coverage Illumina short reads from HG002.
The reads were downloaded from `https://s3-us-west-2.amazonaws.com/human-pangenomics/NHGRI_UCSC_panel/HG002/hpp_HG002_NA24385_son_v1/ILMN/downsampled/`.
We mapped them using `vg giraffe`[@giraffe] on a personalized pangenome[@hapsamp] derived from the HPRC v1.1 Minigraph-Cactus pangenome.
The aligned reads were first stored in GAM, to allow comparing file size and sorting runtimes against GAF.
The GAM file was sorted with `vg gamsort`.
It was then converted to a GAF file using `vg convert`, compressed with plain gzip, and sorted using `vg gamsort` with the new GAF sorting mode described above, with BGZF compression from `bgzip`.
We compared the file sizes and sorting runtimes between both approach.
The query time was also measured on ten thousand regions of the pangenome, specified as node ID ranges. 
These ranges were created by picking a random starting node position on the reference path and walking 50 steps along that path.
The commands and details for this benchmarking is available in the `analysis/readsorting` folder of this paper's repository (see [Code and data availability]).

### Querying GAF files

Instead of indexing on a sequence name and genomic position, we can query on a node interval (see Figure {@fig:index}D).
In HTSlib, `tabix` was modified to disregard any provided sequence name when querying intervals for a GAF file, and to interpret the values typically used for the *start* and *end* position of genomic coordinates as a node ID interval.

Commands to query slices of the pangenome in `vg` were also updated.
The `find` and `chunk` sub-commands now use the updated HTSlib library to extract the appropriate GAF records.
Internally, these commands identify an initial subgraph, which might correspond to, for example, a genomic interval on the reference path provided by the user. 
This subgraph is then extended with additional *context*, allowing a linear-reference-based query to include non-reference regions of the graph.
The amount of context is controlled by the user.
Finally, the paths (historically reads) overlapping the subgraph nodes are extracted.
This last step was updated to be able to extract paths in an indexed GAF file using HTSlib.
It is now possible to extract a slice of an indexed GAF file based on node intervals, coordinates on a reference path, a list of coordinates from a BED file, or a user-provided subgraph, as described in the User Guide in the `analysis` folder of this paper's repository (see [Code and data availability]).

### Projecting annotations into a pangenome

A pangenome represents multiple genomes for which annotations might be available. 
We describe an approach to project annotations relative to a genome into a pangenome.
We updated the `annotate` subcommand from the `vg` toolkit to project regions represented in BED or GFF files into pangenomes and tested it on GBZ-format pangenomes from the HPRC.
Currently, the annotation source genome must be marked as a *reference* in the pangenome, but any genome can be quickly (about a minute in our environment) set to be a reference using the `vg gbwt` subcommand.
Then, `vg annotate` traces out the path in the pangenome graph taken by each input annotation.
The path, represented as an *alignment* record, is then written either in GAM or GAF format.
Sometimes, an annotation's source path might be broken into multiple disjoint parts, because some regions of the individual assemblies are clipped out of the pangenome.
Newly assembled centromeres that challenge current alignment tools, and regions whose alignments would introduce suspiciously large structural variants, are often excluded when pangenomes are constructed from assembled sequences.
Projected annotations are thus also broken up by `vg annotate` as needed when they overlap with breakpoints in the pangenome's representation of their source path.
The name of the annotation record in the output file is taken from the BED file's *name* column, or the *Name* and *ID* GFF fields.

We tested this approach by projecting the gene annotations, repeat annotations, and segmental duplication annotations from each of the 88 assembled haplotypes from the HPRC Year 1 data release into version 1.1 of the GRCh38-based HPRC Minigraph-Cactus pangenome.
The annotations were downloaded from the HPRC public repository at `https://github.com/human-pangenomics/HPP_Year1_Assemblies/tree/main/annotation_index`.
The gene annotations from CAT[@doi:10.1101/gr.233460.117] (GFF files) were downloaded using the URLs in the `Year1_assemblies_v2_genbank_CAT_genes.index` file, the repeat annotations from RepeatMasker (BED files)  from `Year1_assemblies_v2_genbank_Repeat_Masker.index`, segmental duplications (BED files) from `Year1_assemblies_v2_genbank_Seg_Dups.index`, and simple repeats from `Year1_assemblies_v2_genbank_TRF.index`.
A helper script was implemented to prepare the BED files with informative record names.
For example, for repeats from the RepeatMasker annotation, we name each element by its repeat class and repeat family.
The projection of those annotations for each haplotype into the pangenome was automated with a Snakemake workflow available in the `analysis/annotate` folder of this paper's repository (see [Code and data availability]).
<!-- How did we deal with having 88 copies of things when they coincide? -->

### Coverage track from mapped reads

Functional genomics datasets are often visualized as a coverage track.
High coverage in a region might suggest a strong transcription factor binding site or regulatory region. 

We implemented an approach to summarize the coverage of reads across the pangenome into paths along which coverage is consistent (see Figure @fig:meth_cov). 
The coverage in every node is first binned into a few coverage classes, for example representing low, medium and high coverage. 
By default we use 1, 5, and 30 reads as coverage breakpoints to save three bins: 1-5, 5-30, 30+. 
Regions with no coverage are not saved.
Once the coverage is binned, we extend greedily to connected nodes and bins if in the same coverage class.
This extension step produces path through the pangenome with consistent coverage. 
The paths are written in the GAF format, recording the coverage class and the average coverage across the path.

![
**Read coverage bin-and-extend approach to produce coarse-grained coverage tracks.**
In each node (*grey rectangles*), the read coverage at each base is binned using user-defined coverage bins, creating node regions of similar coverage also called bins (*colored rectangles*).
Starting bins (*large green flags*) are selected one at a time, and those bins are extended in both directions until reaching a different-coverage bin (*small red flag*).
](figures/coverage.bin.pangenome.png "Method to make a coverage track from mapped reads"){#fig:meth_cov}

This algorithm was implemented in Python and uses the *libbdsg* module[@libbdsg] to query the pangenome.
It is made available in the public repository of this study in the `analysis/encode` folder of this paper's repository (see [Code and data availability]).

We tested this approach on ATAC-seq datasets from ENCODE[@encode;@encode_portal] on seven tissues: breast epithelium (*ENCFF210QPG*), gastrocnemius medialis (*ENCFF968KAQ*), gastroesophageal sphincter (*ENCFF553SEZ*), Peyer's patch (*ENCFF272PIN*), sigmoid colon (*ENCFF775DFR*), spleen (*ENCFF391IHY*), thyroid gland (*ENCFF579WLS*).
Paired-end short Illumina HiSeq 4000 reads were downloaded from ENCODE and mapped using giraffe[@giraffe] on the HPRC v1.1 Minigraph-Cactus pangenome.
The output to `vg pack` (with `-Q 1` to keep reads with mapping quality of at least 1) was piped to the script implementing the coverage track computation described above.

### Annotating known variants

We implemented an approach to find variants from public databases in the pangenome, producing a GAF file with a pangenome path for each variant.
To accomplish this, we match input variants againt the VCF representation of the pangenome, produced by running `vg deconstruct` on the pangenome file.
This VCF file contains the all variants and their paths through the pangenome in the *AT* *INFO* field.
We look for variants in the pangenome VCF that overlap each imput variant.
When an overlapping variant is found, we extract one of two paths from its *AT* field.
If the input variant is found to be present in the pangenome VCF, we extract the path for the alternate allele.
If the alternate allele of the input variant is not present, but there is still a variant at this position in the pangenome, we extract the path for its reference allele instead.
Variants that are not matched in the pangenome are skipped.
This algorithm was implemented in Python and uses the *libbdsg* module[@libbdsg] internally to get the node size information necessary to write GAF records.

We tested this approach on the GWAS Catalog[@doi:10.1093/nar/gkac1010] and GTEX expression QTLs[@doi:10.1038/ng.2653], and matched them against the HPRC Minigraph-Cactus v1.1 pangenome.
The GWAS Catalog was downloaded from the UCSC Genome Browser's `gwasCatalog` table[@doi:10.1093/nar/gkac1072 ].
The eQTLs from GTEX v8 were downloaded from `https://storage.googleapis.com/adult-gtex/bulk-qtl/v8/single-tissue-cis-qtl/GTEx_Analysis_v8_eQTL.tar`.
The files defining associations between variant/gene pairs were parsed for each tissue.
As before, the output annotations in GAF were bgzipped, sorted and indexed.
We collected statistics about the resulting annotations, and visualized database representation of variants using the Sequence Tube Map.

We implemented another, more straightforward approach, to annotate variants that were genotyped using the pangenome.
Here we simply convert a VCF that contains allele traversal information (*AT* field) to a GAF file representing the alternate allele.
We test this approach by genotyping HG002 from the short-reads Illumina Novaseq reads aligned to the pangenome for the GAF sorting experiment (see [Indexing paths in GAF files]).
Variants were genotyped from the aligned reads using `vg call`.
Of note, `vg call` can also output genotypes directly in GAF format with the `-G` parameter.

The scripts and pipeline to annotate variants is available in the public repository of this study in the `analysis/variants` folder of this paper's repository (see [Code and data availability]).

### Visualization in the Sequence Tube Map

The Sequence Tube Map was developed to interactively explore a pangenome graph, haplotypes traversing it, and reads mapping to it[@tubemap]. 
It internally calls *vg* to extract the relevant slice of the pangenome graph and reads, which previously could only be in the GAM format.
We updated the Sequence Tube Map to accept GAF files that have been sorted, compressed and indexed as in [Indexing paths in GAF files].

<!-- Multi-read-track support is meant to be for the Lancet paper with Rajeeva at NYGC -->
The Sequence Tube Map has also recently added support for displaying multiple layers or "tracks" of haplotypes or reads[@lancet].
Together with the GAF support presented here, this allows the user to add several tracks of annotations from indexed GAF files to the Sequence Tube Map as if they were read files, or alongside actual read files.
The Sequence Tube Map supports assigning a different color or color palette to each track, to facilitate the visualization of different datasets in the same local pangenome region. 
We have extended this functionality with a new option to use an element's opacity to represent an integer score, like a mapping quality or read coverage.
For example, one can now visualize the coding regions of genes and coverage tracks for different cell types from the ENCODE project, as described in [Coverage of seven functional datasets from ENCODE] and highlighted in Figure @fig:tubemap.

### Visualization in Bandage-NG

Bandage-NG[@{https://github.com/asl/BandageNG}], a tool derived from Bandage[@bandage], can read GAF files and visualize its paths through the input graph by coloring the nodes.
Of note is that the offset positions are not taken into account, which means that nodes are colored entirely even if the GAF record only overlaps part of it.
We implemented a wrapper script to facilitate the preprocessing of a subgraph to visualize with Bandage-NG. 
It starts by extracting the subgraph for a region of interest from a full pangenome using `vg chunk`.
This graph is converted to GFA, making sure that the paths are written last for compatibility with Bandage-NG.
The script also extracts annotations relevant to the region of interest from indexed GAF files. 
The extracted GAF files are merged and modified if necessary: path names are de-duplicated and paths are trimmed<!-- Not split??? --> when traversing nodes absent from the extracted subgraph.
These modifications ensure that the final GAF file can be loaded in Bandage-NG.
The outputs of this script can be opened with Bandage-NG for interactive exploration. 
In particular, the *Find path* feature can find nodes corresponding to specific paths, which are either haplotypes or annotations in this pre-processed GFA file.
The user can select a *path* and color its nodes.
The helper script and a tutorial are available at the `analysis/visualization` folder of this paper's repository (see [Code and data availability]).
