## Results

### Sorting and indexing short sequencing reads

Read sorting (see [Indexing paths in GAF files]) was tested on about 30X coverage of Illumina HiSeq paired-end 150bp reads for the HG002 sample.
The plain gzipped GAF file with about 682 million reads was sorted with `vg gamsort` and compressed with `bgzip` in 3h55 using 3.87 hours of CPU time and less than 1 GiB of memory (Table @tbl:readsorting_summary).
Indexing the sorted GAF with `tabix` took 20 minutes using 18.4 minutes of CPU time.
We compared that approach with the existing read sorting implementation in `vg`, which operates on files in the Protobuf-based GAM format[@vg].
Sorting a GAM with the same reads took 12h using 11.87 hours of CPU time and about 7 GiB of memory. 

In addition to being about twice as fast to sort, reads written in the GAF format (and bgzipped) also take about half the disk space (52 GiB vs 108 GiB).
The main reason for this reduced space is that GAF does not save the complete read sequence, only the path through the pangenome and edits to reconstruct it.
GAM files produced by `vg giraffe` can also save additional information as annotations, like the mapping time of each read, which are currently not kept when converting to the GAF format.
Overall, the bgzipped GAF files are half as small and twice as fast to sort for short sequencing reads.

Once a GAF is indexed, extracting a slice of reads works like extracting a slice of an indexed BAM/VCF/BED file for a genomic region, and is more efficient than extracting a slice from an indexed GAM file.
For example, extracting reads for ten thousand random regions in the pangenome took about 0.066 seconds per region to retrieve an average of 1707 reads.
For comparison, the same extraction took an average of 0.816 seconds per region using the GAM format. 

| Format | Time (H:M:S) | Max. memory used (GiB) | File size (Gb) |
|:------:|-------------:|-----------------------:|---------------:|
| GAM    |     12:00:19 |                   6.98 |            108 |
| GAF    |      3:55:13 |                   0.84 |             52 |

Table: Resources used to sort short sequencing reads for a 30x coverage Illumina human dataset.
{#tbl:readsorting_summary}

### Annotation of a human pangenome

To showcase our annotation projection implementation, we projected annotations for all HPRC haplotypes into the HPRC pangenome (see [Projecting annotations into a pangenome]). 
This included genes, segmental duplications, tandem repeats, and repeat annotations. 
`vg annotate` was able to project ~4M gene annotations into the pangenome in ~11 minutes on two cores using 19.9 minutes of CPU time and 21 GiB of RAM.
It projected ~5.5M repeats from RepeatMasker in ~2.9 minutes on two cores using 4.9 minutes of CPU time.
These measures include projection with `vg annotate`, decompression of the input gzipped GAF, and compression of the output GAF with `bgzip`.
We could quickly query these rich annotations with `vg`, and visualize them using tools like the Sequence Tube Map or Bandage-NG.
Using Bandage-NG, we prepared a visualization illustrating a mobile element insertion (Figure {@fig:bandage}).
We also examined a gene annotation using the Sequence Tube Map (Figure {@fig:tubemap}A).

![
**Visualization examples with Bandage-NG**
Example of an AluYa5 transposon insertion (*red*) within the coding sequence of the *PRAMEF4* gene (*blue*).
Both annotations were initially produced at the haplotype level by the Human Pangenome Reference Consortium.
We projected them into the pangenome, indexed them, and queried a small region to visualize with Bandage-NG.
The nodes were colored based on those annotations and loaded as paths by Bandage-NG.
](images/PRAMEF4.AluInsertion.png "Bandage example"){#fig:bandage}

We also matched and annotated variants from the GWAS Catalog[@doi:10.1093/nar/gkac1010](about 660 thousand variants) and expression QTLs from the GTEx catalog[@doi:10.1038/ng.2653] across 49 tissues (on average 1.45 million variants per tissue), using the methods described in [Annotating known variants].
On average, 94% of variants were found in the HPRC pangenome.
The variants files in GAF take only 907 MiB of space and can be queried rapidly for visualization in the Sequence Tube Map or Bandage-NG.
This annotation is showcased in Figure {@fig:tubemap}B.

It is also straightforward to convert genotypes called on variants from the pangenome back to annotation paths. 
This is applicable when the pangenome is used as a source of variants for genotyping in new samples, as when using Pangenie or `vg call`.
To illustrate this use case, we genotyped HG002 using short Illumina reads aligned to the HPRC pangenome with `vg giraffe` and `vg call` (see [Annotating known variants]).
The predicted genotypes were converted to GAF and indexed.
They could be visualized using the Sequence Tube Map, along with the aligned reads. 
This type of annotation can help a developer explore variant calls and aligned reads. 
An example is shown in Figure {@fig:tubemap}C.
On the traditional reference genome, the equivalent would be to load both reads and variants in IGV[@igv].

![
**Visualization examples with the Sequence Tube Map**
**A** Projection of the HPRC gene annotation into the pangenome, here visualizing a coding sequence (CDS) of the *CFD* gene.
There is an insertion (see *insertion node* in the middle), that the reference path (*violet* on top) skips.
Most CDS annotations are similar (*blue* paths), but the CSD from the haplotype with the insertion is split into two parts, highlighted in *green*.
**B.** Projecting known variants into the pangenome. 
*black*: Reference path (GRCh38), *pale colors*: other human haplotypes, *reds*: GWAS Catalog, *blues*: eQTLs across tissues (GTEx).
**C.** Aligned reads and genotype calls.
*yellow*/*green*: annotation paths from `vg call` genotypes. *reds*/*blues*: short sequencing reads.
**D.** ATAC-seq coverage track for seven tissues from the ENCODE project. 
The reference path (*violet*) and other HPRC haplotypes (*greys*) are shown on top and traverse two variation sites.
ATAC-seq coverage annotations for the different tissues are shown in different *colors* with different opacities representing coverage level.
The thyroid annotation (*blue*) is more opaque than others, suggesting that this region is more open in that tissue.
The *red* annotations at the bottom show the position of exons from the *TPO* gene, a thyroid-specific gene.
](images/tubemap.examples.png "Tubemap examples"){#fig:tubemap}

### Coverage of seven functional datasets from ENCODE

We aligned ENCODE ATAC-seq datasets from 7 cell types[@encode] to the draft human pangenome to produce coverage tracks as indexed GAF files (see [Coverage track from mapped reads]). 
On average across cell types, there were about 475 thousand paths representing high read coverage, which were, on average, 2.6 nodes and 104.8 bases long. 
An average of 63 thousand paths with high ATAC-seq read coverage traversed three or more nodes (see Table @tbl:coverage_summary).
Because the graph node length used in non-variable regions is longer than the path length, these paths visiting three or more nodes are likely in regions of the pangenome with variation.

| Dataset                    |   Paths | Average bases | Average nodes | Traversing >2 nodes |
|:---------------------------|--------:|--------------:|--------------:|--------------------:|
| Breast epithelium          | 570,155 |        109.80 |          2.68 |      65,095 (11.4%) |
| Gastrocnemius medialis     | 342,830 |         94.47 |          2.43 |      53,595 (15.6%) |
| Gastroesophageal sphincter | 555,618 |        115.69 |          2.73 |        72,407 (13%) |
| Peyerâ€™s patch              | 270,094 |         97.88 |          2.62 |      43,591 (16.1%) |
| Sigmoid colon              | 678,380 |        115.10 |          2.69 |      82,906 (12.2%) |
| Spleen                     | 531,910 |        104.97 |          2.66 |      67,153 (12.6%) |
| Thyroid gland              | 377,157 |         95.63 |          2.41 |        56,441 (15%) |

Table: High coverage tracks from seven functional datasets on the HPRC pangenome. For each sample, the table shows how many paths had a mean coverage of at least ten reads and how long they were.
{#tbl:coverage_summary}

It took on average 7 CPU-hours to map the reads to the pangenome using VG Giraffe, and 2.8 CPU-hours to produce the coverage tracks.
Sorting, compressing and indexing them took only 0.23 CPU-hours, on average.
Table @tbl:coverage_benchmark compiles the runtime and memory usage for each step across all samples.

Figure {@fig:tubemap}D shows an example of a region, visualized with our extended version of the Sequence Tube Map (see [Visualization in the Sequence Tube Map]), near an exon of the *TPO* gene. 
The exon annotations came from projecting the HPRC gene annotations described above into the pangenome. 
The coverage track for the ATAC-seq dataset shows that this region is only opened in the thyroid, which is consistent with the thyroid-specific role of *TPO*.
By integrating these two external sources of annotations (gene annotation and ATAC-seq coverage), we can visualize them in the context of the genomic variation in the pangenome.

<!-- `example of what we could look for and describe`{.red} -->
<!-- Fig @fig:cov_examples shows examples of those tracks visualized using the Sequence Tube Map. -->
<!-- In Fig ??a, the promoter of the *??* gene is seen to be open in the ?? cell type.  -->
<!-- Thanks to the pangenomic view, we see differential coverage of the functional tracks across the variants in the region. -->
<!-- For instance, we notice a small insertion-deletion (indel) where the alternate allele is only covered by 2 reads, while the reference allele is covered by more than 30 reads. -->
<!-- Fig ??b highlights a structural variant, a ??bp insertion, that is highly covered by ATAC-seq in several cell types. -->
<!-- The RepeatMasker annotation in this region, also extracted from an indexed GAF file, flags this insertion as a ?? transposable element. -->
<!-- ?? can indeed attract TF?? that lead to open chromatin ?REF?. -->

<!-- ![ -->
<!-- **Coverage tracks visualized interactively using the Sequence Tube Map.** -->
<!-- a) Promoter... b) Structural variant.... -->
<!-- ](figures/wide.png "Wide image"){#fig:cov_examples} -->

### Efficient subgraph extraction and speed-up for the Sequence Tube Map

A new subgraph extraction was implemented using three tabix-indexed files: a GAF file with haplotype information, a BED file associating coordinates on the haplotypes to nodes, and a TSV file associating each node to its sequence. 
The three index files were produced in about 1h40m starting from the gzipped GFA file of the HPRC Minigraph-Cactus v1.1 pangenome.

Extracting a subgraph on 20 random regions of 100 bp, 1 Kbp, or 10 Kbp took about 0.5s per query, compared to 33s using `vg chunk` (Figure {@fig:pgchunking}).
Of note, as other tabix-indexed files, the new pangenome files can be queried remotely through HTTPS/FTP.
For example, we extracted the same regions using files hosted in a server in the United States, from a computer in France, in 15s on average. 
In this case, the small TBI files are automatically downloaded locally, if needed, which added about additional 20 second to the first query.

![
**Subgraph extraction using tabix-indexed files**.
Benchmark performed on the HPRC Minigraph-Cactus v1.1 human pangenome.
**Left** Size of the index files. 
**Right** Time to query 20 random regions of varying sizes (y-axis), using the tabix-indexed files (`chunkix` method) or `vg chunk` from the GBZ file.
The `chunkix` method also works on remote files, here through HTTPS, and is labelled here as `chunkix_remote`. 
In that case the small TBI files must be downloaded locally which increase the time for the first query (highlighted with a diamond).
](images/pgchunking.fig.png "Subgraph extraction"){#fig:pgchunking}

On disk, the three index files totaled 4 GiB: 2.4 GiB for the "haplotype" bgzipped GAF file, 1.1 GiB for the "node sequence" bgzipped TSV file, and 527 MiB for the "haplotype position" bgzipped BED file (Figure {@fig:pgchunking}).
These index files are 2.75 smaller than input gzipped GFA file (11 GiB) and 30% larger than the GBZ file (3.1 GiB).

Thanks to their integration in the sequenceTubeMap, it only takes about a second to visualize a region in the sequenceTubeMap when using these new index files, compared to more than 30s before.

