## Discussion

The tools and applications described here streamline the production of annotations in the GAF format, their indexing, and efficient querying.

First, these techniques will help integrate additional information into existing or future pangenomic tools.
They will simplify the integration of information like gene or repeat annotations, known variants, or more generally other functional annotations, in pangenomic analysis.
These analyses will be able to work with those annotations in light of underlying genomic variation described in the pangenome.
Visualization is a specific example where the users typically want to include several layers of annotations, including some of their own design.
Tools to visualize pangenomes will greatly benefit from a simple pipeline to create or load pangenomic annotations.

Second, this work is critical to allow the community to make, share and reuse annotations in the pangenome space.
The GAF format is already used by multiple independent tools, although mostly for read mapping.
Thanks to our work, GAF files can now be indexed and queried efficiently, like BED, VCF, or GFF on a linear reference genome.
This will prove useful for those working with GAF-format read mappings, and also enables the foirmat to be used for large annotation sets.
We showcase how annotations on a linear genome can be projected into the pangenome and stored in GAF format.
We believe it could become the de facto format to represent annotations on the pangenome, thus accelerating adoption of the new pangenomic paradigm by the broader genomics field.
<!-- Although informative annotations could be analyzed already, the current approach has some limitations. -->
<!-- - Simplistic handling of clipped paths. -->

The indexing scheme relies on integer node IDs, compacted relative to their position in the pangenome.
If node IDs are not integers, it adds a layer of complexity for the user as they will need to convert them to integers and keep track of the translation between original pangenome and new pangenome.
Both `vg` and `odgi` offer ways to convert pangenome to a new pangenome with compact integer node IDs, but tool support for the resulting translation information is uneven and no standard format exists to store it.

Our implementation of GAF-based annotation is designed with short paths in mind, where the user wants to extract the full annotation paths for all annotations in a region.
This is convenient when working with short reads, gene annotations, and most genomic repeats, for example.
Larger genomic regions, like chromosome bands, large assembled contigs, or large segmental duplications, can still be represented and manipulated, but might be less practical to work with.
With the current implementation, the full annotation will be extracted when querying any overlapping region.
The extracted annotations are not clipped to the specified range, and no tooling support is provided for finding the part or parts of each annotation record that are actually relevant to the query.
To address this use case, we plan to implement an extraction mode where output annotations are trimmed to keep only parts overlapping the queried subgraph.

The na√Øve metadata integration is another limitation.
Indeed, the input annotation information from BED and GFF3 files is currently reduced to a single string label.
In our applications, those labels contain, for example, a gene name and haplotype names, or a read coverage bin and average coverage value, all in one label.
This was easy to implement and sufficient for now because there are no tools that handle more advanced metadata organization.
For many annotations, it would be useful to keep the metadata better organized, so that the user can access/use it within visualization tools.
The different metadata fields could be saved using GAF tags that can be added at the end of each GAF record.

Our approach to converting annotations from a linear genome to the pangenome assumes that we have annotations of the different haplotypes in the pangenome.
There is still no clear solution to lift annotations from one reference/haplotype to other haplotypes in the pangenome; reanalysis/reannotation of each haplotype is generally required.
The homology information embedded in the pangenome can sometimes be used to propagate annotations from one haplotype to others more easily. 
This strategy is used by annotation tools like CAT[@doi:10.1101/gr.233460.117], and improves over reanalysis of each haplotype from scratch.
In the future, these techniques might help propagate other types of annotations across the pangenome more efficiently than by reanalyzing the raw data from scratch on each haplotype.
`something about odgi paths/untangle maybe`{.red}

Our GAF-based approach also highlights the limitations of the existing tool support for GAF files.
Some tools, like Bandage-NG or GfaViz, require manual pre-processing, such as extracting a subgraph and integrating the annotations as embedded paths.
The Sequence Tube Map can now handle indexed bgzipped GAF files, but the query time for large pangenomes remains long in practice.
Integrating annotation metadata into its interface as a first-class data type will also require a significant amount of development.
Overall, we stress the need for visualization tools that can efficiently lay out and organize many paths through a pangenome.

We have shown that the GAF format, thanks to new tools for its efficient manipulation, could offer a path for the future of annotations in pangenome graphs.
While it provides an important building block, it is clear that more needs to be done to make it a useful solution for the community.


## Code and data availability

`vg` is available at [https://github.com/vgteam/vg](https://github.com/vgteam/vg).
Bandage-NG is available at [https://github.com/asl/BandageNG](https://github.com/asl/BandageNG).
The Sequence Tube Map is hosted at [https://github.com/vgteam/sequenceTubemap](https://github.com/vgteam/sequenceTubemap)

The analysis presented in this manuscript is documented in the repository at [https://github.com/jmonlong/manu-vggafannot](https://github.com/jmonlong/manu-vggafannot)[@repo].
It contains scripts used to prepare the annotation files, commands and automated pipelines used to annotate them in the pangenome, and helper scripts to summarize the output files.

The annotations of the HPRC v1.1 pangenome were deposited on Zenodo at `??ZENODO_LINK??`{.red}.
This includes gene annotations, repeats from RepeatMasker, simple repeats and segmental duplications.
A coverage track for the seven ENCODE ATAC-seq samples is also available in this repository.

## Acknowledgments

We would like the thank the ENCODE consortium and the laboratory of Michael Snyder for making the ATAC-seq datasets available.

## Author contribution statement

AMN, GH, JM, TTY, and EG contributed code to the vg toolkit (*annotate*, *pack*, *call* subcommands) with support from BP.
AMN integrated new code to HTSlib.
AMN and DC implemented new features in the Sequence Tube Map.
SD selected and analyzed the ATAC-seq datasets from ENCODE.
JM conceived the study, analyzed data, and drafted the manuscript.
All authors contributed in reviewing the text and content of the manuscript.
