Scripts and workflows to reproduce the analysis in the manuscript

- [`annotate`](annotate): Projecting HPRC haplotype annotations into the pangenome using `vg annotate`.
- [`encode`](encode): Making coverage track for ATAC-seq datasets from ENCODE.
- [`readsorting`](readsorting): Sorting short sequencing reads mapped to the pangenome.
- [`variants`](variants): Projecting known SNPs/indels (GWAS catalog, eQTLs, ...) into the pangenome.
- [`visualization`](visualization): Visualizing some example.
- [`data.md`](data.md): information about the data produced in this manuscript and [deposited on Zenodo](https://doi.org/10.5281/zenodo.13904205)


Below are general commands used by those scripts and workflows to prepare, sort, compress, index, and query annotations in GAF.

## Getting started with GAF indexing

- GAF specs: https://github.com/lh3/gfatools/blob/master/doc/rGFA.md#the-graph-alignment-format-gaf

*Soon...*
